internal/fs/type_cache_test.go

```diff
@@ -0,0 +1,421 @@
+
+////////////////////////////////////////////////////////////////////////
+// Common
+////////////////////////////////////////////////////////////////////////
+
+type typeCacheTestCommon struct {
+	fsTest
+}
+
+var (
+	// The following should be configured for different tests
+	// differently inside SetUpTestSuite as these need to
+	// set for mount itself.
+
+	// ttlInSeconds is equivalent of metadata-cache:ttl-secs in config-file.
+	ttlInSeconds int64
+
+	// typeCacheMaxSizeMb is equivalent of metadata-cache:type-cache-max-entries in config-file.
+	typeCacheMaxSizeMb int
+)
+
+func (t *typeCacheTestCommon) SetUpTestSuite() {
+	t.serverCfg.MountConfig = config.NewMountConfig()
+	t.serverCfg.MountConfig.MetadataCacheConfig = config.MetadataCacheConfig{
+		TypeCacheMaxSizeMB: typeCacheMaxSizeMb,
+		TtlInSeconds:       ttlInSeconds,
+	}
+
+	// Fill server-cfg from mount-config.
+	func(mountConfig *config.MountConfig, serverCfg *gcsfusefs.ServerConfig) {
+		serverCfg.DirTypeCacheTTL = mount.MetadataCacheTTL(mount.DefaultStatOrTypeCacheTTL, mount.DefaultStatOrTypeCacheTTL,
+			mountConfig.TtlInSeconds)
+		serverCfg.InodeAttributeCacheTTL = serverCfg.DirTypeCacheTTL
+		// We can add more logic here to fill other fileds in serverCfg
+		// from mountConfig here as needed.
+	}(t.serverCfg.MountConfig, &t.serverCfg)
+
+	// Call through.
+	t.fsTest.SetUpTestSuite()
+}
+
+////////////////////////////////////////////////////////////////////////
+// Specific test classes
+////////////////////////////////////////////////////////////////////////
+
+type TypeCacheTestWithMaxSize1MB struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithMaxSize1MB) SetUpTestSuite() {
+	ttlInSeconds = 30
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithZeroCapacity struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithZeroCapacity) SetUpTestSuite() {
+	ttlInSeconds = 30
+	typeCacheMaxSizeMb = 0
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithZeroTTL struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithZeroTTL) SetUpTestSuite() {
+	ttlInSeconds = 0
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithInfiniteTTL struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithInfiniteTTL) SetUpTestSuite() {
+	ttlInSeconds = -1
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+func init() {
+	RegisterTestSuite(&TypeCacheTestWithMaxSize1MB{})
+	RegisterTestSuite(&TypeCacheTestWithZeroCapacity{})
+	RegisterTestSuite(&TypeCacheTestWithZeroTTL{})
+	RegisterTestSuite(&TypeCacheTestWithInfiniteTTL{})
+}
+
+// //////////////////////////////////////////////////////////////////////
+// helpers
+// //////////////////////////////////////////////////////////////////////
+func (t *typeCacheTestCommon) testNoInsertion() {
+	const name1 = "foo"
```

Comment:
I am seeing, these logic repeated:
(a) creation of "foo" object.
(b) Stat call to validate.
(c) Creation of "foo/" object.
(d) Stat call to validate.

Can we create utility method for the above, so that we can remove the duplicacy at every places?

Reply:
done.